vsutil.py

```diff
@@ -176,3 +195,75 @@ def is_image(filename: str, /) -> bool:
+        -> vs.VideoNode:
+    """
+    A bit depth converter only using core.resize and Format.replace.
+    By default, outputs FLOAT sample type for 32 bit and INTEGER for anything else.
+
+    :param bitdepth:    Desired bits_per_sample of output clip.
+    :param sample_type: Desired sample_type of output clip. Allows overriding default FLOAT/INTEGER behavior. Accepts
+                        vapoursynth.SampleType enums INTEGER and FLOAT or their values, [0, 1].
+    :param range:       Output pixel range (defaults to input clip's range). See `Range`.
+    :param range_in:    Input pixel range (defaults to input clip's range). See `Range`.
+    :param dither_type: Dithering algorithm. Allows overriding default dithering behavior. See `Dither`.
+        Defaults to Floyd-Steinberg error diffusion when downsampling, converting between ranges, or upsampling full
+        range input. Defaults to 'none', or round to nearest, otherwise.
+
+    :return: Converted clip with desired bit depth and sample type. ColorFamily will be same as input.
+    """
+    def readable_enums(enum, module: str = 'vsutil') -> list:
+        """
+        Returns a readable error message listing all possible values in `module.enum`.
+        """
+        return [module + "." + str(i) for i in enum] + [i.value for i in enum]
+
+    # enum mapping and error checking
+    if sample_type is not None:
+        try:
+            sample_type = vs.SampleType(sample_type)
+        except ValueError:
+            raise ValueError(f'depth: sample_type must be in {readable_enums(vs.SampleType, str(vs.__name__))}.') from None
+    if range is not None:
+        try:
+            range = Range(range)
+        except ValueError:
+            raise ValueError(f'depth: range must be in {readable_enums(Range)}.') from None
+    if range_in is not None:
+        try:
+            range_in = Range(range_in)
+        except ValueError:
+            raise ValueError(f'depth: range_in must be in {readable_enums(Range)}.') from None
+    if dither_type is not None:
+        try:
+            dither_type = Dither(dither_type)
+        except ValueError:
+            raise ValueError(f'depth: dither_type must be in {readable_enums(Dither)}.') from None
+
+    curr_depth = get_depth(clip)
+    if sample_type is None:
+        if bitdepth == 32:
+            sample_type = vs.FLOAT
+        else:
+            sample_type = vs.INTEGER
+
+    if (curr_depth, clip.format.sample_type, range_in) == (bitdepth, sample_type, range):
+        return clip
+
+    if dither_type is None:
+        if range_in != range or range_in == 1:  # thanks @Frechdachs for explaining this: 'you need dithering when raising
+            dither_type = Dither.ERROR_DIFFUSION  # the bitdepth of full range [or] converting between full and limited'
+        else:
+            dither_type = Dither.ERROR_DIFFUSION if curr_depth > bitdepth and sample_type == vs.INTEGER else Dither.NONE
```

Comment:
There are only two possible outcomes here (none or error diffusion), but we have three code paths.
We could express it as a single if/else:
```py
if range_in != range or range_in == Range.FULL or (curr_depth > bitdepth and sample_type == vs.INTEGER):
  Dither.ERROR_DIFFUSION
else:
  Dither.NONE

Reply:
```py 
dither_type = (dither_type 
               or (Dither.ERROR_DIFFUSION
                   if (range_in != range 
                       or range_in == Range.FULL 
                       or (curr_depth > bitdepth and sample_type == vs.INTEGER))
                   else Dither.NONE))
```
yeah, I'm flexing

Reply:
I thought about that as well, but in the end it almost felt less readable to me, despite probably being more pythonic.
EIther way, if you’re going with that, you can use `fallback(dither_type, ...)` instead of `(dither_type or ...)`.

Reply:
```py
should_dither = (range_in != range
                 or range_in == Range.FULL
                 or (curr_depth > bitdepth and sample_type == vs.INTEGER)
dither_type = fallback(dither_type, Dither.ERROR_DIFFUSION if should_dither else Dither.NONE)
```
this is less flexy

Reply:
Addressed in a9c672116db536ada8c16c06843b89932266a234.